compiler/rustc_hir_typeck/src/opaque_types.rs

```diff
@@ -0,0 +1,29 @@
+use super::FnCtxt;
+impl<'tcx> FnCtxt<'_, 'tcx> {
+    /// After cloning the opaque types storage during writeback to compute the
+    /// defining uses, we may in theory add further uses of the opaque type.
+    ///
+    /// Silently ignoring them is dangerous and could result in ICE or even in
+    /// unsoundness, so we make sure we catch such cases here. There's currently
+    /// no known code where this actually happens, even with the new solver which
+    /// does normalize types in writeback after cloning the opaque type storage.
+    ///
+    /// FIXME(@lcnr): I believe this should be possible in theory and would like
+    /// an actual test here. After playing around with this for an hour, I wasn't
+    /// able to do anything which didn't already try to normalize the opaque before
+    /// then, either allowing compilation to succeed or causing an ambiguity error.
+    pub(super) fn detect_opaque_types_added_during_writeback(&self) {
+        let num_entries = self.checked_opaque_type_storage_entries.take().unwrap();
+        let newly_added_opaque_types: Vec<_> =
+            self.inner.borrow_mut().opaque_types().opaque_types_added_since(num_entries).collect();
+        for (key, hidden_type) in newly_added_opaque_types {
+            let opaque_type_string = self.tcx.def_path_str(key.def_id);
+            let msg = format!("unexpected cyclic definition of `{opaque_type_string}`");
+            self.dcx()
+                .struct_span_err(hidden_type.span, msg)
+                .with_note("please open a GitHub issue: https://github.com/rust-lang/rust/issues")
+                .emit();
```

Comment:
This will just  cause an error-class message, won't it? Maybe make this an ICE so we can find this with crater or fuzzers? :thinking: